A syntax for Modifiers

goldfirere opened the pull request:
The [proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0370-modifiers.rst) has been accepted; the following discussion is mostly of historic interest.

---
This proposes a new syntax for *modifiers*, a prefix word or expression that modifies the way a bit of syntax behaves. The only modifier available right now is the multiplicity of an arrow, but #242 will want a new modifier, and the proposal outlines other ways this new feature might be used.

The primary motivation behind this is to create a palatable way to combine linear types with unsaturated type families, and is the first time I feel happy about the way these two features will interact.

[Rendered](https://github.com/goldfirere/ghc-proposals/blob/modifiers/proposals/0000-modifiers.rst)

goldfirere committed:
Propose *modifiers* for GHC

goldfirere committed:
Formatting

goldfirere committed:
Link to PR

jvanbruegge commented:
I like this, it cleans up this interaction quite well. I also support not defaulting to Multiplicity as I for example would use Matchability more than Multiplicity (and it just feels wrong to arbitrarily default to one)

Just one question: The `m` is a normal type variable, isn't it? So instead of a kind signature directly with the modifier anywhere else should be fine too, is it? So it is just that the modifier by itself does not allow ghc to infer a kind and thus the kind needs to be user supplied.

int-index commented:
Yes, it would be helpful to show that one `:: Multiplicity` annotation is sufficient:

```
map :: forall (p :: Multiplicity) a b. (a %p-> b) -> [a] %p-> [b]
```

jvanbruegge commented:
And I  guess map would probably written more like this:
```
map :: forall {p :: Multiplicity} a b. (a %p -> b) -> [a] %p -> [b]
```
(With an inferred argument)

goldfirere committed:
Modifiers in more locations, à la linear types

goldfirere commented:
Prodded by @monoidal, I added support for modifiers in more places, to keep up with the linear types proposal. I've also added the `map` example above.

simonpj commented:
In the first bullet of "Costs" you imply that you have to say `Int %(m :: Multiplicity) -> Bool`.  But you don't, as the last example in Examples shows.   Confusing!

`data Tagged (%nominal t) a = Tagged a`.  I didn't understand this one.  Where is `nominal` (presumably a type variable) bound? A and how is this decoration reflected in the type it decorates?

There may a tricky bit implementation-wise.  We don't know whether the modifier is a multplicity or a matchability thing until after kind inference has completed.  Another reason for kind-checking to return a HsType not a Type.

Presumably the kind of the modifier must be one of a fixed collection.  No modifier polymoprhism!  Say this.

Modifiers are always types, right, even in terms?  `prefix% atype`.

goldfirere committed:
Clarifications from GitHub discussion

goldfirere commented:
I've added a few clarifications based on @simonpj's comment.

> Modifiers are always types, right, even in terms?

I interpret this question as: "what parser and namespace are used in modifiers?" For now, yes, they're types. But there's no reason this has to remain true, as we continue to find uses for this new syntax.

aspiwack commented:
I don't understand why the `f5` example would parse (even to be rejected). I don't think it follows from your specification, which I read as allowing a single `%mod` to the left of an arrow.

int-index commented:
@aspiwack In EBNF, `{ ... }` means repetition, i.e. "zero or more". So `{modifier}` accepts a sequence of modifiers.

aspiwack commented:
Ah, indeed. Thanks.

kcsongor commented:
I'm strongly in favour if this proposal! (somehow missed it until now, thanks for writing it up)
Concerning #242, there is also a need to annotate `forall`s, e.g.:
```haskell
type Foo :: forall k. @U k
type family Foo :: k where ...
```

could/should this be handled by rule 3 in your proposal?

```haskell
type Foo :: forall (%U k). k
```

The current version of #242 uses a single annotation for the whole telescope of forall-bound variables, but more fine-grained control might be better.

goldfirere commented:
This has been quiet. I'd like to submit to the committee @nomeata . I would normally say that it should be considered alongside #242, but I'm shepherding that, and so it seems improper to also shepherd this one.

serras commented:
I really think this is the way to go, since more and more proposals are going to add syntax. Let me ask a few questions about the proposal:
- I feel like @simonpj's question about inference have not been fully answered. The proposal says that a constraint is emitted, but having a variable lying around may negatively affect how constraint solving proceeds.
- Could we introduce a generic way to say which is the "default" of each modifier. I mean, when we parse `a -> b`, somehow we have to know that a lone arrow means matchability X, linearity Y, and so on.
- Do we need to decide on syntax for the term language, or would it be better to focus simply on the type level?

goldfirere commented:
> 
>     * I feel like @simonpj's question about inference have not been fully answered. The proposal says that a constraint is emitted, but having a variable lying around may negatively affect how constraint solving proceeds.

Yes. I'm unsure about this. I can't seem to quite figure out how it would all play out.

> 
>     * Could we introduce a generic way to say which is the "default" of each modifier. I mean, when we parse `a -> b`, somehow we have to know that a lone arrow means matchability X, linearity Y, and so on.

I don't see this as a property of the modifier, but rather of the `->`. I could imagine writing e.g. `default Matchability M` or `default Multiplicity One` or some such, but I think that's separate from this proposal.

> 
>     * Do we need to decide on syntax for the term language, or would it be better to focus simply on the type level?

I would like to universally reserve prefix-`%` for modifiers, meaning that no other proposal should take this space, unless that proposal is compatible with a modifier interpretation. Beyond reserving term-level syntax, I don't need this proposal to affect the term level.

serras commented:
I am recommending acceptance on this proposal. I'll leave up to the Committee to decide whether the prefix should also be reserved at the term-level.

gridaphobe commented on proposals/0000-modifiers.rst:
```diff
@@ -0,0 +1,233 @@
+10. During constraint generation, let an occurrence ``%(ty)``, where ``ty :: ki``,
+    emit a constraint ``Modifier ki``.
+
+11. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
```
The idea of constraint resolution changing the type of a neighboring syntactic element feels really strange to me. And it's very unclear how we would extend this to user-defined modifiers without a user-definable semantics for "running" a modifier, which I think is a must for a feature like this (see other existing mechanisms like `ANN` pragmas, the various plugins we support, even template haskell and quasiquoters). I think I'd be happier with this proposal if it specified a user-definable mechanism and then showed how we could use it to solve the multiplicity/matchability problem.

goldfirere replied:
Yes, I've grown uncomfortable about this, too. Modification of the proposal to appear shortly.

goldfirere commented on proposals/0000-modifiers.rst:
```diff
@@ -0,0 +1,233 @@
+
+6. With ``-XModifiers``, introduce postfix modifier syntax on record field declarations as follows::
+     
+     fielddecl ::= vars '::' (type | '!' atype) {modifier}
```
This is meant to be compatible with the current syntax for linear types. I'm open to other placements.

gridaphobe replied:
One possibility would be to follow Haddock's precedent and have one syntax for modifying the next syntactic element, and another for modifying the previous element, but then we'd have to steal more syntax.

Also, I should note that the second sentence of the proposal says:

> Modifiers somehow change the meaning of the **next** token.

goldfirere commented on proposals/0000-modifiers.rst:
```diff
@@ -0,0 +1,233 @@
+  tantalizing.
+  
+* These modifiers recall Java's `Annotations <https://en.wikipedia.org/wiki/Java_annotation>`_
+  mechanism, which were a direct inspiration.
```
I think it depends on what we want the semantics of a pragma to be. I like the idea that pragmas should be out-of-band instructions to the compiler. Inlining, rewrite rules, warning enable/disable, and deprecation (among others) fit nicely here. On the other hand, annotations that change the typing or execution (beyond performance) of the program are really not out-of-band: they are an essential part of the program. These, in my opinion, should not be pragmas -- instead, these should be modifiers. The `ANN` mechanism you mention here thus seems similar to, but not the same as, the idea of modifiers.

gridaphobe replied:
I think changing program semantics or not may be a useful distinction to make, but this proposal already blurs the lines by suggesting `inline` and `oneShot` as candidates for future modifiers. Neither changes the semantics of your program, only how GHC optimizes it. And it's also not super clear where `ANN` pragmas fall. They don't change the semantics on their own, but seem designed to aid plugins that may well use them to change your program's semantics!

Here's another example, I've long wanted to give LiquidHaskell refinements a more first-class feeling by teaching GHC about them to some extent. We considered using `ANN` pragmas at one point but it didn't quite work for some technical reason (I think it was surmountable, we just didn't follow through..). But would they fit better as modifiers? From GHC's perspective refinements don't change the semantics of your program (at least not at the moment), but they are an essential piece from the programmer's perspective.

goldfirere replied:
These are all really good points -- the line is blurrier than I realized. And you're right about `oneShot` and `inline`: these really should be pragmas, not modifiers. It's just that they are now magic functions, which seems worse than either.

For what it's worth, I would say that an LH annotation would be a modifier, because it changes the compile-time semantics of the program (taking a program that might have otherwise been accepted to be rejected).

serras commented:
@goldfirere given that changes are expected here, I'll move this proposal back to the revision state. Looking forward to see those changes, I think that this proposal -- in whatever shape ends up being -- is really needed for the upcoming changes in the language.

goldfirere committed:
Remove the Modifier class, simplifying proposal.

goldfirere commented:
I've uploaded my changes, removing `Modifier` and simplifying the proposal somewhat. Really, this is all about reserving syntactic space, and I agree that the overloading mechanism I tried to build was a step too far.

Given the technical nature of this one and the general lack of commentary, I'd like to send this pack to the committee, @serras.

gridaphobe commented on proposals/0000-modifiers.rst:
```diff
@@ -0,0 +1,214 @@
+
+     type     ::= btype [ {modifier} -> type ]
+     prefix%  ::= '%'    -- only in prefix position
+     modifier ::= prefix% atype
```
It still feels strange to say that a modifier is a type. It sounds more like a function `Type -> Type` (or `Expr -> Expr` or `Pat -> Pat` in the other cases), but now that sounds a lot like a Template Haskell splice. Should modifiers maybe be a nicer syntax for running TH splices?

goldfirere replied:
That's an interesting way to phrase them, but no. If we imagine that TH supports a larger syntax than Haskell, we could also imagine implementing lots of features via TH that have no syntax in "pure" Haskell. But I would rather keep TH and Haskell aligned with each other -- and to avoid needing the TH machinery to, say, specify a linear type.

And a modifier is not a type. It's parsed/renamed as one, but then interpreted separately.

gridaphobe replied:
> But I would rather keep TH and Haskell aligned with each other -- and to avoid needing the TH machinery to, say, specify a linear type.

Agreed, but I assumed there would be a modifier-free way of specifying things like linear types (possibly horribly ugly) and that modifiers would act more as a shorthand for the full incantation, which feels more like TH.

gridaphobe commented on proposals/0000-modifiers.rst:
```diff
@@ -0,0 +1,214 @@
+  discussions may have enjoyed having the modifier option. For example,
+  this might have been used instead of ``type role`` for role annotations:
+  ``data Tagged (%Nominal t) a = Tagged a``. Or it might have been an
+  alternative for ``-XDerivingStrategies``.
```
Another possibility that just came to mind, should we have levity modifiers?

goldfirere replied:
We could have levity modifiers, but as levity is nicely captured by the kind system, there seems to be only weak motivation to introduce a new mechanism for them.

gridaphobe replied:
Another feature that in retrospect seems like a modifier: strictness annotations.

goldfirere committed:
A few clarifications

aspiwack commented:
There seem to be a similar flavour from this new variant of the proposal than the _fully settled_ idea previously discussed at https://github.com/ghc-proposals/ghc-proposals/pull/216#issuecomment-598744392 . Which makes sense to me.

nomeata commented:
> There seems to be a similar flavour from this new variant of the proposal than the _fully settled_ idea previously discussed at [#216 (comment)](https://github.com/ghc-proposals/ghc-proposals/pull/216#issuecomment-598744392) . Which makes sense to me.

So if we had had these ~annotations~ modifiers earlier, I wonder if that might have swayed us to the builder-design for local do, using the  `%builder do` synax? But probably otiose to muse about that.

simonpj commented:
> This proposal means that Int %m -> Bool, acceptable today as a multiplicity-polymorphic function, would be rejected.

Would it? What about
```
f :: forall (m :: Multiplicity). Int %m -> Bool
```
surely that's clear enough? Why can't we accept it?

> Future modifiers will also seamlessly work with existing ones, where order is not expected to matter (though that would be up to other proposals to spell out).

Can we specify, up front, that order doesn't matter?

**Syntax**: the form `btype [ {modifier} -> type ]` seems to allow one or zero modifiers but not two.  Or do I misunderstand the braces?  Spell it out please

simonpj commented:
Generally I'm in favour, but I'd mark this as experimental syntax.  We might end up wanting to change it.  But the only we can find out is to try it.

kcsongor commented:
I see that there's no implementation plan outlined in the proposal. May I volunteer to implement it? Given that unsaturated type families will be the primary beneficiary of this feature in the short term, I think it would make sense to bundle the implementation effort.

goldfirere commented:
Yes -- I think implementing this within the implementation for #242 would be ideal. Thanks!

goldfirere commented:
@simonpj 

> Syntax: the form `btype [ {modifier} -> type ]` seems to allow one or zero modifiers but not two. Or do I misunderstand the braces? Spell it out please

Directly above this production, the proposal says

> recall that we use braces to denote "zero or more"

Does that specify the grammar precisely?

simonpj commented:
>     recall that we use braces to denote "zero or more"

Well, I missed that.  Sorry.  Maybe put the parenthesis *after* the BNF block rather than before?

goldfirere commented:
I'm inclined to leave this part alone -- I think introducing notation before using it makes more sense than the other way around.

Is there anything else blocking a merge?

googleson78 commented:
I'm interested in implementing this but the best I can do currently is a "a couple of hours on some weekends" sort of commitment. I'd also appreciate someone to "mentor" me, mostly on design points. 

@kcsongor are you still working/planning to work on this, or can I potentially pick it up at some point?

goldfirere commented:
Happy to help work out a design. And the time you're offering here is a lot more than we've got, so it's very welcome. Thank you! If you want to go ahead, make a GHC ticket, tag me, and then it's also probably best to email, just to make sure it gets my attention.

Icelandjack commented:
This could be used to give field or constructor specific deriving behaviour, as a continuation of a [prior idea](https://www.reddit.com/r/haskell/comments/pbq9rn/via_fields_finer_granularity_in_deriving/). Specifying `%(Via Sum Int) Int` hints that we want derive using the implementation for `Sum Int`.

```haskell
data XY = MkXY
  { x :: %(Via Sum Int) Int
  , y :: %(Via Sum Int) Int }
  deriving stock Generic
  deriving (Semigroup, Monoid) via Generically XY
```

The `!`-field modifier could then be viewed as shorthand for this sort of `%Bang` modifier.

googleson78 commented:
I've also been constantly thinking about what @Icelandjack said - a lot of other languages have some pragma that says things like "omit this field from the json rep" or "this field is going to be the db key" and modifiers have the potential to fill in this exact role verry well if we expose modifier info to `Generic` and TH.
